vhost-device-spi/src/spi.rs

```diff
@@ -0,0 +1,1055 @@
+    }
+
+    fn set_max_speed_hz(&self, max_speed_hz: u32) -> Result<()> {
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret = unsafe {
+            ioctl_with_ref(
+                &self.file.as_raw_fd(),
+                SPI_IOC_WR_MAX_SPEED_HZ(),
+                &max_speed_hz,
+            )
+        };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("set_max_speed_hz", IoError::last()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn bits_per_word(&self) -> Result<u8> {
+        let mut bpw: u8 = 0;
+
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret = unsafe {
+            ioctl_with_mut_ref(&self.file.as_raw_fd(), SPI_IOC_RD_BITS_PER_WORD(), &mut bpw)
+        };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("bits_per_word", IoError::last()))
+        } else {
+            Ok(bpw)
+        }
+    }
+
+    fn set_bits_per_word(&self, bpw: u8) -> Result<()> {
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret =
+            unsafe { ioctl_with_ref(&self.file.as_raw_fd(), SPI_IOC_WR_BITS_PER_WORD(), &bpw) };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("set_bits_per_word", IoError::last()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn mode(&self) -> Result<u32> {
+        let mut mode: u32 = 0;
+
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret =
+            unsafe { ioctl_with_mut_ref(&self.file.as_raw_fd(), SPI_IOC_RD_MODE32(), &mut mode) };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("mode", IoError::last()))
+        } else {
+            Ok(mode)
+        }
+    }
+
+    fn set_mode(&self, mode: u32) -> Result<()> {
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret = unsafe { ioctl_with_ref(&self.file.as_raw_fd(), SPI_IOC_WR_MODE32(), &mode) };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("set_mode", IoError::last()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn rdwr(&self, msgs: &mut [SpiIocTransfer]) -> Result<()> {
+        let len = msgs.len();
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret = unsafe {
+            ioctl_with_mut_ptr(
+                &self.file.as_raw_fd(),
+                spi_ioc_message(len as u32),
+                msgs.as_mut_ptr(),
+            )
+        };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("rdwr", IoError::last()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn detect_supported_features(&self) -> Result<VirtioSpiConfig> {
+        // supported cs_max_number 1
+        // can't set cs timing from userland in Linux, reserve cs timing as 0
+        // cs_change_supported always enabled, cause Linux can handle this in software
+        // max_word_delay_ns reserved as 0, also can't set from userland
+
+        //detect max_speed_hz
```

Comment:
Please always use the same style, for example leaving a space from `// ` as you did a few lines above.

Reply:
Also in general, I suggest you use comments only to explain things that are not clear, here I think the name of the function is self-explanatory (correctly), so this comment seems superfluous.

Reply:
I agree, and will remove the superfluous comments

Reply:
The aim of this comment is to separate `detect_supported_features` function into some parts, and each corresponds with the interface with the hardware. So how about reserving these comments?

Reply:
Sure, just make sure to keep the style consistent. Adding to the comments to explain that each part corresponds to an interface would make the comments clear too

Reply:
OK